third_party/terraform/resources/resource_dataflow_job.go

```diff
@@ -191,51 +191,70 @@ func resourceDataflowJobDelete(d *schema.ResourceData, meta interface{}) error {
 	}
 
 	id := d.Id()
-	requestedState, err := mapOnDelete(d.Get("on_delete").(string))
+
+	requestedState, err := resourceDataflowJobMapRequestedState(d.Get("on_delete").(string))
 	if err != nil {
 		return err
 	}
-	for _, ok := dataflowTerminalStatesMap[d.Get("state").(string)]; !ok; _, ok = dataflowTerminalStatesMap[d.Get("state").(string)] {
+
+	// Retry updating the state while the job is not ready to be canceled/drained.
+	err = resource.Retry(time.Minute*time.Duration(15), func() *resource.RetryError {
+		// To terminate a dataflow job, we update the job with a requested
+		// terminal state.
 		job := &dataflow.Job{
 			RequestedState: requestedState,
 		}
 
-		_, err = updateJob(config, project, region, id, job)
-		if err != nil {
-			if gerr, err_ok := err.(*googleapi.Error); !err_ok {
+		_, updateErr := resourceDataflowJobUpdateJob(config, project, region, id, job)
+		if updateErr != nil {
+			gerr, isGoogleErr := err.(*googleapi.Error)
+			if !isGoogleErr {
 				// If we have an error and it's not a google-specific error, we should go ahead and return.
-				return err
-			} else if err_ok && strings.Contains(gerr.Message, "not yet ready for canceling") {
-				// We'll sleep below to wait for the job to be ready to cancel.
-			} else {
-				return err
+				return resource.NonRetryableError(err)
+			}
+
+			if strings.Contains(gerr.Message, "not yet ready for canceling") {
+				// Retry cancelling job if it's not ready.
+				// Sleep to avoid hitting update quota with repeated attempts.
+				time.Sleep(5 * time.Second)
```

Comment:
I think you should be able to control this with `MinTimeout` instead of manual sleeps.

Reply:
MinTimeout like operation MinTimeout? I think we had this here before because dataflow doesn't use operations :/